src/core/components/pin.js

```diff
@@ -0,0 +1,438 @@
+              if (err) { return cb(err) }
+              // found all objects, we can add the pin
+              return cb(null, key)
+            })
+          } else {
+            if (recursivePins.has(key)) {
+              // recursive supersedes direct, can't have both
+              return cb(new Error(`${key} already pinned recursively`))
+            }
+            if (directPins.has(key)) {
+              // already directly pinned
+              return cb(null, key)
+            }
+            // make sure we have the object
+            dag.get(new CID(multihash), (err) => {
+              if (err) { return cb(err) }
+              // found the object, we can add the pin
+              return cb(null, key)
+            })
+          }
+        }), (err, results) => {
+          if (err) { return callback(err) }
+          // update the pin sets in memory
+          if (recursive) {
+            results.forEach(key => {
+              // recursive pin should replace direct pin
+              directPins.delete(key) // TODO questionable
+              recursivePins.add(key)
+            })
+          } else {
+            results.forEach(key => directPins.add(key))
+          }
+          // persist updated pin sets to datastore
+          pin.flush((err, root) => {
+            if (err) { return callback(err) }
+            self.log(`Added pins: ${results}`)
+            return callback(null, results.map(key => ({hash: key})))
+          })
+        })
+      })
+    }),
+
+    rm: promisify((paths, options, callback) => {
+      let recursive = true
+      if (typeof options === 'function') {
+        callback = options
+      } else if (options && options.recursive === false) {
+        recursive = false
+      }
+      callback = once(callback)
+      resolvePaths(self, paths, (err, mhs) => {
+        if (err) { return callback(err) }
+        // verify that each hash can be unpinned
+        series(mhs.map(multihash => cb => {
+          pin.isPinnedWithType(multihash, pin.types.all, (err, res) => {
+            if (err) { return cb(err) }
+            const { pinned, reason } = res
+            const key = toB58String(multihash)
+            if (!pinned) {
+              return cb(new Error(`${key} is not pinned`))
+            }
+
+            switch (reason) {
+              case (pin.types.recursive):
+                if (recursive) {
+                  return cb(null, key)
+                } else {
+                  return cb(new Error(`${key} is pinned recursively`))
+                }
+              case (pin.types.direct):
+                return cb(null, key)
+              default:
+                return cb(new Error(
+                  `${key} is pinned indirectly under ${reason}`
+                ))
+            }
+          })
+        }), (err, results) => {
+          if (err) { return callback(err) }
+
+          // update the pin sets in memory
+          results.forEach(key => {
+            if (recursive) {
```

Comment:
Could be simplified to

```
if (recursive && recursivePins.has(key)) {
    recursivePins.delete(key)
} else {
    directPins.delete(key)
}
```
Though I'm not sure if that's easier to read/understand, so I leave it up to you if you if you want to leave it as it currently is or not.

Reply:
Nope, it is easier! On it.